[rpc] add the worker IDs outside of addSendRpcBackward to ensure they are
recorded in all cases

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30914 [rpc] add the worker IDs outside of addSendRpcBackward to ensure they are recorded in all cases**

When tensors don't require grad, we don't call `addSendRpcBackward`, where we record known workerIDs to clean up the dist autograd context later. But since  https://github.com/pytorch/pytorch/pull/29781, we always include the autograd context ID in RPCs, even if tensors do not require grad. So, it could be possible that we don't release the contexts on some nodes.

This can contribute to OOMs since the contexts will not be cleaned up in this case, which can be checking by running the unit test without this patch. We can fix this issue by moving the `addKnownWorkerIds`  call to the `getMessageWithAutograd` function.

Differential Revision: [D18869191](https://our.internmc.facebook.com/intern/diff/D18869191/)

rohan-varma committed:
[rpc] add the worker IDs outside of addSendRpcBackward to ensure they are
recorded in all cases

When tensors don't require grad, we don't call `addSendRpcBackward`, where we record known workerIDs to clean up the dist autograd context later. But since D18496709, we always include the autograd context ID in RPCs, even if tensors do not require grad. So, it could be possible that we don't release the contexts on some nodes.

This can contribute to OOMs since the contexts will not be cleaned up in this case, which can be checking by running the unit test without this patch. We can fix this issue by moving the `addKnownWorkerIds`  call to the `getMessageWithAutograd` function.

Differential Revision: [D18869191](https://our.internmc.facebook.com/intern/diff/D18869191/)

[ghstack-poisoned]

pritamdamania87 commented on test/dist_autograd_test.py:
nit: would be nice to add some docs explaining what issue the unit test is catching

rohan-varma committed:
Update on "[rpc] add the worker IDs outside of addSendRpcBackward to ensure they are
recorded in all cases"


When tensors don't require grad, we don't call `addSendRpcBackward`, where we record known workerIDs to clean up the dist autograd context later. But since  https://github.com/pytorch/pytorch/pull/29781, we always include the autograd context ID in RPCs, even if tensors do not require grad. So, it could be possible that we don't release the contexts on some nodes.

This can contribute to OOMs since the contexts will not be cleaned up in this case, which can be checking by running the unit test without this patch. We can fix this issue by moving the `addKnownWorkerIds`  call to the `getMessageWithAutograd` function.

Differential Revision: [D18869191](https://our.internmc.facebook.com/intern/diff/D18869191/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] add the worker IDs outside of addSendRpcBackward to ensure they are
recorded in all cases"

recorded in all cases**

When tensors don't require grad, we don't call `addSendRpcBackward`, where we record known workerIDs to clean up the dist autograd context later. But since  https://github.com/pytorch/pytorch/pull/29781, we always include the autograd context ID in RPCs, even if tensors do not require grad. So, it could be possible that we don't release the contexts on some nodes.

This can contribute to OOMs since the contexts will not be cleaned up in this case, which can be checking by running the unit test without this patch. We can fix this issue by moving the `addKnownWorkerIds`  call to the `getMessageWithAutograd` function.

Differential Revision: [D18869191](https://our.internmc.facebook.com/intern/diff/D18869191/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] add the worker IDs outside of addSendRpcBackward to ensure they are
recorded in all cases"

recorded in all cases**
recorded in all cases**

When tensors don't require grad, we don't call `addSendRpcBackward`, where we record known workerIDs to clean up the dist autograd context later. But since  https://github.com/pytorch/pytorch/pull/29781, we always include the autograd context ID in RPCs, even if tensors do not require grad. So, it could be possible that we don't release the contexts on some nodes.

This can contribute to OOMs since the contexts will not be cleaned up in this case, which can be checking by running the unit test without this patch. We can fix this issue by moving the `addKnownWorkerIds`  call to the `getMessageWithAutograd` function.

Differential Revision: [D18869191](https://our.internmc.facebook.com/intern/diff/D18869191/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] add the worker IDs outside of addSendRpcBackward to ensure they are
recorded in all cases"

recorded in all cases**

When tensors don't require grad, we don't call `addSendRpcBackward`, where we record known workerIDs to clean up the dist autograd context later. But since  https://github.com/pytorch/pytorch/pull/29781, we always include the autograd context ID in RPCs, even if tensors do not require grad. So, it could be possible that we don't release the contexts on some nodes.

This can contribute to OOMs since the contexts will not be cleaned up in this case, which can be checking by running the unit test without this patch. We can fix this issue by moving the `addKnownWorkerIds`  call to the `getMessageWithAutograd` function.

Differential Revision: [D18869191](https://our.internmc.facebook.com/intern/diff/D18869191/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] add the worker IDs outside of addSendRpcBackward to ensure they are
recorded in all cases"

recorded in all cases**
recorded in all cases**

When tensors don't require grad, we don't call `addSendRpcBackward`, where we record known workerIDs to clean up the dist autograd context later. But since  https://github.com/pytorch/pytorch/pull/29781, we always include the autograd context ID in RPCs, even if tensors do not require grad. So, it could be possible that we don't release the contexts on some nodes.

This can contribute to OOMs since the contexts will not be cleaned up in this case, which can be checking by running the unit test without this patch. We can fix this issue by moving the `addKnownWorkerIds`  call to the `getMessageWithAutograd` function.

Differential Revision: [D18869191](https://our.internmc.facebook.com/intern/diff/D18869191/)

[ghstack-poisoned]
